Sched cache resync

anfernee opened the pull request:
**What this PR does / why we need it**:  Scheduler cache comparer
    
    A debug tool that collects resources from api server and compares it
    with the scheduler cache. It currently only compares the node list, but
    it should be easy to extend. The compare is triggered by signal USER2,
    by doing
    
      kill -12 ${SCHED_PID}
    
    The compare result goes to scheduler log.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Towards #60860

**Special notes for your reviewer**: @bsalamat 

**Release note**:
```release-note
None
```

Commit:
Support snapshotting a scheduler cache

Towards #60860

Commit:
Scheduler cache comparer

A debug tool that collects resources from api server and compares it
with the scheduler cache. It currently only compares the node list, but
it should be easy to extend. The compare is triggered by signal USER2,
by doing

  kill -12 ${SCHED_PID}

The compare result goes to scheduler log.

Towards #60860

anfernee commented:
In order to make it simple, it's signal triggered now. but I am open to suggestions like option, periodical ticker, etc.

anfernee commented on pkg/scheduler/schedulercache/cache.go:
```diff
@@ -80,6 +80,33 @@ func newSchedulerCache(ttl, period time.Duration, stop <-chan struct{}) *schedul
 	}
 }
 
+// Snapshot takes a snapshot of the current schedulerCache.
```
this has performance impact, now only used for debug purpose.

resouer replied:
Why don't add this NOTE in the comment here?

anfernee replied:
agreed it would be nice to add in comment.

bsalamat commented:
/assign
I took a quick look and the direction is good. Please add comparision for Pods as well.

Commit:
Add cache comparison for pods and pdbs

anfernee commented:
@bsalamat pods added. more comparison can be added too. but at this point, I am not what's more valuable. right now, mismatching pod, node, pdb will be logged.

anfernee commented:
/retest

bsalamat reviewed (commented):
Thanks, @anfernee!

bsalamat commented on pkg/scheduler/factory/cache_comparer.go:
```diff
@@ -0,0 +1,154 @@
+		glog.Warningf("cache mismatch: missed nodes: %s; redundant nodes: %s", missed, redundant)
+	}
+
+	if missed, redundant := c.ComparePods(pods, snapshot.Nodes); len(missed)+len(redundant) != 0 {
```
This part of the PR is not correct and will create false positives. Scheduler keeps assigned pods (pods with PodSpec.NodeName != "") in its cache. Unassigned pods live in scheduler's scheduling queue. For now it is OK to ignore unassigned pods. The code in its current form compares all pods against assigned ones in the scheduler cache. So, if there are pending pods in the cluster, we will see warning messages.

anfernee replied:
I am not aware of that.. Let me fix it and rebase this PR.

anfernee replied:
Updated @bsalamat

Commit:
Pod comparer should count pods in scheduling queue

Pods in scheduler cache contains both the scheduled pods and those not
scheduled yet in scheduling queue. This commit adds the second group of
pods into consideration while comparing the cache.

anfernee commented:
/retest

bsalamat commented:
/lgtm

Thanks, @anfernee!
I think we should probably get this in 1.10 as well. This may help find an important issue that is causing a lot of trouble for many users.

anfernee commented:
@bsalamat sure. if it's helpful. it touches a lot but overall not risky, it is not affecting the main workflow.

bsalamat commented:
I don't think it is urgent enough to be merged during the current code freeze. We can wait until after the code freeze and then backport it to 1.10 for the next dot release of 1.10.

anfernee commented:
i am fine either way.

wenjiaswe commented:
cc @roycaihw @wenjiaswe

resouer commented on pkg/scheduler/factory/signal.go:
```diff
@@ -0,0 +1,25 @@
+
+// compareSignal is the signal to trigger cache compare. For non-windows
+// environment it's SIGUSR2.
+var compareSignal = syscall.SIGUSR2
```
I think we will enable this in test-infra, how will test-infra expect to trigger it may I ask?

anfernee replied:
you mean test this in e2e test? no, it's for whoever interested in triggering the comparison to debug. Nothing would happen except some logs are appended.

resouer replied:
I saw it was mentioned in the issue description: 
> We will probably enable it in our tests and in clusters where customers see stale cache issues

Anyway this question will not block this PR. I will re-lgtm it when CI passed.

bsalamat replied:
@resouer, Your point is valid. In fact, I also chatted with @anfernee on Slack about periodic triggering of this or enabling it in tests. I believe we should try this feature and make sure that we do not see a lot of false positives. We should also measure its performance overhead in larger tests to ensure that enabling it in tests won't cause timeouts. Once it is refined, we may want to enable it in certain tests.

resouer replied:
Got it, thanks for explaining.

resouer commented:
/lgtm

Thanks for quick fix the nits.

Random-Liu commented:
I guess this changes broke the test `[sig-scheduling] SchedulerPredicates [Serial] validates that taints-tolerations is respected if not matching`.
https://k8s-testgrid.appspot.com/google-gce#gci-gce-serial

The test started failing after this change list https://github.com/kubernetes/kubernetes/compare/3e5596321...025439988. And this PR seems most relevant.

@anfernee @bsalamat

resouer commented:
And I believe the break is caused by: https://github.com/kubernetes/kubernetes/pull/61089

Random-Liu commented:
@resouer Thanks for looking into this!
